[tensorexpr] Add missing python bindings for NNC Stmts

huiguoo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#55570 [tensorexpr] Add missing python bindings for NNC Stmts**



Differential Revision: [D27634987](https://our.internmc.facebook.com/intern/diff/D27634987)

huiguoo committed:
[tensorexpr] Add NNC conv tests in python; Add required python bindings for NNC Stmts

[ghstack-poisoned]

huiguoo commented on test/test_tensorexpr_pybind.py:
```diff
@@ -40,6 +41,109 @@ def compute(i):
+
+            def get_dim_args(dims):
+                dim_args = []
+                for dim in dims:
+                    dim_args.append(torch._C._te.DimArg(dim, 'i' + str(len(dim_args))))
+                return dim_args
+
+            Pimage = torch._C._te.Placeholder('image', dtype, [NN, CC, HI, WI])
+            Pweight = torch._C._te.Placeholder('weight', dtype, [KK, CC, RR, SS])
+            Pout = torch._C._te.Placeholder('conv', dtype, [NN, KK, HH, WW])
+
+            dint = torch._C._te.Dtype.Int
+            n, k, h, w, c, r, s = [torch._C._te.VarHandle(s, dint) for s in ["n", "k", "h", "w", "c", "r", "s"]]
+            stmt = loop(n, NN,
+                    loop(k, KK,
+                        loop(h, HH,
+                            loop(w, WW,
+                                loop(c, CC,
+                                    loop(r, RR,
+                                        loop(s, SS,
+                                            Pout.store(
+                                                [n, k, h, w],
+                                                Pout.load([n, k, h, w]) +
+                                                Pimage.load([n, c, h+r, w+s]) *
+                                                Pweight.load([k, c, r, s])
+                                                )))))))
+                    )
+
+            loopnest = torch._C._te.LoopNest(stmt, [Pout.buf()])
+            loopnest.prepare_for_codegen()
+            stmt = torch._C._te.simplify(loopnest.root_stmt())
+            codegen = torch._C._te.construct_codegen('llvm', stmt, [torch._C._te.BufferArg(x) for x in [Pimage, Pweight, Pout]])
+
+            image = torch.ones(N, C, H, W)
+            weight = torch.ones(K, C, R, S)
+            ref= F.conv2d(image, weight)
+            out = torch.zeros_like(ref)
+            codegen.call([image.float(), weight.float(), out.float()])
+            torch.testing.assert_allclose(out, ref)
+
```
Not sure if it's proper to add the tests here. We already have a conv test written in cpp. @ZolotukhinM

ZolotukhinM reviewed (commented):
Overall looks like a good addition, but please address the linter errors first - it's hard to review the PR appropriately now, the linter is too loud :)

huiguoo committed:
Update on "[tensorexpr] Add missing python bindings for NNC Stmts"




Differential Revision: [D27634987](https://our.internmc.facebook.com/intern/diff/D27634987)

[ghstack-poisoned]

huiguoo commented:
@ZolotukhinM Removed the conv tests - they are bulky and we have already have one conv test in cpp. btw, is there a way to put this back for review? So hard to find the key.

ZolotukhinM reviewed (approved):
That looks good, thanks! But you will need to rebase to a more recent master: I recently removed masks from Loads and Stores, so you will need to update your PR accordingly before landing it.

> btw, is there a way to put this back for review? So hard to find the key.

There is a "re-request review" button in the reviewers list for each person who has reviewed before.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/tensorexpr_init.cpp:
```diff
@@ -127,7 +127,26 @@ void initTensorExprBindings(PyObject* module) {
+            return self.storeWithMask(args, val, mask);
+          })
+      .def("buf", [](Placeholder& self) { return BufHandle(self.data()); })
+      .def("data", [](Placeholder& self) { return self.data(); });
```
I don't think we ever return raw pointers in python?

ZolotukhinM commented on torch/csrc/jit/tensorexpr/tensorexpr_init.cpp:
```diff
@@ -232,12 +275,57 @@ void initTensorExprBindings(PyObject* module) {
+             const ExprHandle& start,
+             const ExprHandle& stop,
+             Stmt* body,
+             const LoopOptions& loop_options) {
```
Have we bound `LoopOptions` to python?

huiguoo committed:
Update on "[tensorexpr] Add missing python bindings for NNC Stmts"




Differential Revision: [D27634987](https://our.internmc.facebook.com/intern/diff/D27634987)

[ghstack-poisoned]

huiguoo committed:
Update on "[tensorexpr] Add missing python bindings for NNC Stmts"




Differential Revision: [D27634987](https://our.internmc.facebook.com/intern/diff/D27634987)

[ghstack-poisoned]
